prow/jira/fakejira/fake.go

```diff
@@ -116,44 +131,73 @@ func (f *FakeClient) CreateIssueLink(link *jira.IssueLink) error {
 	if err != nil {
 		return fmt.Errorf("failed to get outward link issue: %v", err)
 	}
-	outward.Fields.IssueLinks = append(outward.Fields.IssueLinks, link)
+	// links in an issue struct do not include the short definition of the the issue they are part of.
+	// This behavior is used by the external jira-lifecycle hook plugin to identify which direction a clone happened,
+	// so it needs to be replicated in the fake client
+	linkForOutward := *link
+	linkForOutward.OutwardIssue = nil
+	outward.Fields.IssueLinks = append(outward.Fields.IssueLinks, &linkForOutward)
 	inward, err := f.GetIssue(link.InwardIssue.ID)
 	if err != nil {
 		return fmt.Errorf("failed to get inward link issue: %v", err)
 	}
-	inward.Fields.IssueLinks = append(inward.Fields.IssueLinks, link)
+	linkForInward := *link
+	linkForInward.InwardIssue = nil
+	inward.Fields.IssueLinks = append(inward.Fields.IssueLinks, &linkForInward)
 	f.IssueLinks = append(f.IssueLinks, link)
 	return nil
 }
 
 func (f *FakeClient) CloneIssue(issue *jira.Issue) (*jira.Issue, error) {
-	// create deep copy of parent so we can modify key and id for child
-	data, err := json.Marshal(issue)
-	if err != nil {
-		return nil, err
-	}
-	issueCopy := &jira.Issue{}
-	err = json.Unmarshal(data, issueCopy)
-	if err != nil {
-		return nil, err
-	}
-	// set ID and Key to unused id and key
-	f.updateIssueIDAndKey(issueCopy)
-	// run generic cloning function
-	return jiraclient.CloneIssue(f, issueCopy)
+	return jiraclient.CloneIssue(f, issue)
 }
 
 func (f *FakeClient) CreateIssue(issue *jira.Issue) (*jira.Issue, error) {
```

Comment:
What, exactly, is this helping? Having to build a fake Jira sounds like testing is happening at the wrong level. See also: how this repo has not built a fake GitHub, Bugzilla, S3, etc.

Reply:
What do you mean by this repo not building a fake github or bugzilla? There is a fakegithub and a fake bugzilla client that are used heavily by unit testing in plugins. This fake jira client isn't much different.

Reply:
Those are simple in-memory maps that we add to / delete from using the methods. They do not encode provider-specific implementation details like `("issue.fields.project must be set to create new issue")`.